packages/hint/src/lib/types/html.ts

```diff
@@ -0,0 +1,120 @@
+import * as parse5 from 'parse5';
+import * as htmlparser2Adapter from 'parse5-htmlparser2-tree-adapter';
+import * as cssSelect from 'css-select';
+
+import {
+    INamedNodeMap,
+    ProblemLocation
+} from '../types';
+
+type Attrib = {
+    [key: string]: string;
+};
+
+export type ParsedHTMLElement = Element & {
+    attribs: Attrib;
+    sourceCodeLocation: parse5.ElementLocation;
+    parent: ParsedHTMLElement | null;
+}
+
+export class HTMLElement {
+    public ownerDocument?: HTMLDocument;
+
+    private _element: ParsedHTMLElement;
+
+    public constructor(element: ParsedHTMLElement | HTMLElement, ownerDocument?: HTMLDocument) {
+        this._element = element instanceof HTMLElement ? element._element : element;
+        this.ownerDocument = ownerDocument;
+    }
+
+    public get attributes(): INamedNodeMap {
+        const x = this._element.attribs;
+
+        return Object.entries(x).map((y) => {
+            return {
+                name: y[0],
+                value: y[1] as string
+            };
+        });
+    }
+
+    public get nodeName(): string {
+        return this._element.tagName;
+    }
+
+    public getAttribute(attribute: string): string | null {
+        const attrib = this._element.attribs[attribute];
+        const value = typeof attrib !== 'undefined' ? attrib : null;
+
+        return value;
+    }
+
+    public getLocation(): ProblemLocation {
+        const location = this._element.sourceCodeLocation;
+
+        return {
+            column: location.startCol,
+            line: location.startLine
+        };
+    }
+
+    public isSame(element: HTMLElement): boolean {
+        return this._element === element._element;
+    }
+
+    public outerHTML(): string {
+        /*
+         * Until parse5 support outerHTML
+         * (https://github.com/inikulin/parse5/issues/230)
+         * we need to use this workaround.
+         * https://github.com/inikulin/parse5/issues/118
+         *
+         * The problem with this workaround will modify the
+         * parentElement and parentNode of the element, so we
+         * need to restore it before return the outerHTML.
+         */
+        const fragment = htmlparser2Adapter.createDocumentFragment();
+        const parent = this._element.parent;
+
+        htmlparser2Adapter.appendChild(fragment, this._element);
```

Comment:
Does this potentially alter order relative to siblings as well when we restore it?

Reply:
Nope.

Reply:
I will take another look, but IIRC, this only alter the parent element.

Reply:
At least for now, the only change the parent property: (prev is always empty, because we are using a new fragment everytime.
```js
const appendChild = (exports.appendChild = function(parentNode, newNode) {
    const prev = parentNode.children[parentNode.children.length - 1];

    if (prev) {
        prev.next = newNode;
        newNode.prev = prev;
    }

    parentNode.children.push(newNode);
    newNode.parent = parentNode;
});
```

Anyways, I think I'm going to add code to restore `parent`, `prev` and `next`, just in case this code change in the future.